File: kdcreader_android/android/src/main/kotlin/koamtac/kdc/sdk/plugins/delegate/KReaderDelegate.kt
package koamtac.kdc.sdk.plugins.delegate

import io.flutter.plugin.common.MethodChannel
import koamtac.kdc.sdk.KDCReader
import koamtac.kdc.sdk.plugins.KConstants
import java.lang.ref.WeakReference

abstract class KReaderDelegate: KDelegate {
    lateinit var wrReader: WeakReference<KDCReader>

    fun isConnected(reader: KDCReader?): Boolean {
        return reader?.IsConnected() ?: false;
    }

    fun sendErrorResult(err: Map<String, Any>, result: MethodChannel.Result) {
        sendErrorResult(err, null, result);
    }

    fun sendErrorResult(err: Map<String, Any>, detail: Any?, result: MethodChannel.Result) {
        val code = err.get(KConstants.ErrorCode.KEY_CODE) as String;
        val message = err.get(KConstants.ErrorCode.KEY_MESSAGE) as String;
        result.error(code, message, detail)
    }
}

File: kdcreader_ios/ios/Classes/delegate/KReaderDelegate.swift
//
//  KReaderDelegate.swift
//  kdcreader_ios
//
//  Created by koamtac on 3/19/24.
//

import Flutter
import Foundation

//protocol KReaderDelegate: AnyObject {
//    func isSupported(_ name: String) -> Bool
//
//    func run(_ call: FlutterMethodCall, result: @escaping FlutterResult)
//}

class KReaderDelegate: KDelegate {
    let reader: KDCReader
    let converter: KConverter
    
    var SUPPORTED_ACTIONS = [String]()
    
    init(reader: KDCReader, converter: KConverter) {
        self.reader = reader
        self.converter = converter
    }
    
    func isSupported(_ name: String) -> Bool {
        return SUPPORTED_ACTIONS.contains(name)
    }

    func run(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        return false
    }
    
    func isConnected() -> Bool {
        return reader.isConnected()
    }
    
    func sendErrorResult(_ err:Dictionary<String, Any>, detail: Any? = nil, result: @escaping FlutterResult) {
        guard let code = err[KConstants.ErrorCode.Code] as? String,
              let message = err[KConstants.ErrorCode.Message] as? String else {
            return
        }
        
        result(FlutterError(code: code, message: message, details: detail))
    }
}
